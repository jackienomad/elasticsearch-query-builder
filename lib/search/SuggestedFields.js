export class SuggestedFields {
  constructor(mappedFields) {
    this.mappedFields = mappedFields

    if (!this.mappedFields) {
      throw new Error(
        'SuggestedFields class requires a field mapping to be provided'
      )
    }
  }

  buildQuery(keyword, startDate, endDate) {
    // check if keyword is blank
    if (!keyword) throw new Error('Keyword must not be blank')

    const indexes = Object.keys(this.mappedFields).map(key => {
      const { index, field } = this.mappedFields[key]
      return {
        index: `${index}`,
        type: `${field.toLowerCase()}`
      }
    })

    const query = {
      query: {
        bool: {
          must: {
            match_phrase: {
              Suggestion: `${keyword}`
            }
          },
          must_not: {
            match: {
              'Suggestion.untouched': `${keyword}`
            }
          },
          filter: {
            range: {
              ArrivalDate: {
                gte: +startDate,
                lte: +endDate
              }
            }
          }
        }
      },
      size: 0
    }

    for (let i = 0, len = indexes.length; i < len; i++) {
      indexes.splice(i * 2 + 1, 0, query)
    }

    return {
      body: indexes
    }
  }

  formatResponse(esResponse) {
    const { responses } = esResponse

    let fields = Object.keys(this.mappedFields)
    fields = fields.filter((item, index) => {
      const {
        hits: { total }
      } = responses[index]
      return +total > 0
    })

    return fields
  }
}

export default SuggestedFields
