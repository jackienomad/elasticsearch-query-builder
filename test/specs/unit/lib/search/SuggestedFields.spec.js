import SuggestedFields from 'qb/@/search/SuggestedFields'
import path from 'path'
import {
  esResponse,
  esResponseWithZeroShipper
} from './dataset/suggestedFieldElasticSearchResponse'

describe('SuggestedFields class', () => {
  const mappedFields = {
    Consignee: {
      field: 'Suggestion',
      index: 'us_import_consignee',
      type: 'query',
      search_field: 'Consignee'
    },
    Shipper: {
      field: 'Suggestion',
      index: 'us_import_shipper',
      type: 'query',
      search_field: 'Shipper'
    },
    MarksAndNumbers: {
      field: 'Suggestion',
      index: 'us_import_marks_and_numbers',
      type: 'query',
      search_field: 'MarksAndNumbers'
    },
    ProductKeywords: {
      field: 'Suggestion',
      index: 'us_import_product_keywords',
      type: 'query',
      search_field: 'ProductDescription'
    }
  }

  describe('constructor', () => {
    it('can be instantiated with a field mapping', () => {
      expect(() => {
        const suggestedFields = new SuggestedFields(mappedFields)
      }).not.toThrowError()
    })

    it('should throw an error if a field mapping is not provided', () => {
      expect(() => {
        const suggestedFields = new SuggestedFields()
      }).toThrowErrorMatchingInlineSnapshot(
        `"SuggestedFields class requires a field mapping to be provided"`
      )
    })
  })

  describe('buildQuery', () => {
    it('should return an msearch query', () => {
      const keyword = 'apple'
      const startDate = 1542250403
      const endDate = 1542250403

      const suggestedFields = new SuggestedFields(mappedFields)
      const query = suggestedFields.buildQuery(keyword, startDate, endDate)
      expect(query).toMatchSnapshot()
    })

    it('should throw error on empty keyword', () => {
      const keyword = ''
      const startDate = 1542250403
      const endDate = 1542250403

      const suggestedFields = new SuggestedFields(mappedFields)
      expect(() => {
        suggestedFields.buildQuery(keyword, startDate, endDate)
      }).toThrowErrorMatchingInlineSnapshot(`"Keyword must not be blank"`)
    })
  })

  describe('formatResponse', () => {
    it('should be able to return formatted elasticsearch response', async () => {
      const suggestedFields = new SuggestedFields(mappedFields)
      const input = esResponse
      const output = suggestedFields.formatResponse(input)
      const expectedOutput = [
        'Consignee',
        'Shipper',
        'MarksAndNumbers',
        'ProductKeywords'
      ]
      expect(output).toEqual(expectedOutput)
    })

    it('should be able remove field on zero hits', async () => {
      const suggestedFields = new SuggestedFields(mappedFields)
      const input = esResponseWithZeroShipper
      const output = suggestedFields.formatResponse(input)
      const expectedOutput = ['Consignee', 'MarksAndNumbers', 'ProductKeywords']
      expect(output).toEqual(expectedOutput)
    })
  })
})
