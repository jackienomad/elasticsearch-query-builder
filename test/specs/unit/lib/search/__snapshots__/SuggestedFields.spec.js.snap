// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestedFields class buildQuery should return an msearch query 1`] = `
Object {
  "body": Array [
    Object {
      "index": "us_import_consignee",
      "type": "suggestion",
    },
    Object {
      "query": Object {
        "bool": Object {
          "filter": Object {
            "range": Object {
              "ArrivalDate": Object {
                "gte": 1542250403,
                "lte": 1542250403,
              },
            },
          },
          "must": Object {
            "match_phrase": Object {
              "Suggestion": "apple",
            },
          },
          "must_not": Object {
            "match": Object {
              "Suggestion.untouched": "apple",
            },
          },
        },
      },
      "size": 0,
    },
    Object {
      "index": "us_import_shipper",
      "type": "suggestion",
    },
    Object {
      "query": Object {
        "bool": Object {
          "filter": Object {
            "range": Object {
              "ArrivalDate": Object {
                "gte": 1542250403,
                "lte": 1542250403,
              },
            },
          },
          "must": Object {
            "match_phrase": Object {
              "Suggestion": "apple",
            },
          },
          "must_not": Object {
            "match": Object {
              "Suggestion.untouched": "apple",
            },
          },
        },
      },
      "size": 0,
    },
    Object {
      "index": "us_import_marks_and_numbers",
      "type": "suggestion",
    },
    Object {
      "query": Object {
        "bool": Object {
          "filter": Object {
            "range": Object {
              "ArrivalDate": Object {
                "gte": 1542250403,
                "lte": 1542250403,
              },
            },
          },
          "must": Object {
            "match_phrase": Object {
              "Suggestion": "apple",
            },
          },
          "must_not": Object {
            "match": Object {
              "Suggestion.untouched": "apple",
            },
          },
        },
      },
      "size": 0,
    },
    Object {
      "index": "us_import_product_keywords",
      "type": "suggestion",
    },
    Object {
      "query": Object {
        "bool": Object {
          "filter": Object {
            "range": Object {
              "ArrivalDate": Object {
                "gte": 1542250403,
                "lte": 1542250403,
              },
            },
          },
          "must": Object {
            "match_phrase": Object {
              "Suggestion": "apple",
            },
          },
          "must_not": Object {
            "match": Object {
              "Suggestion.untouched": "apple",
            },
          },
        },
      },
      "size": 0,
    },
  ],
}
`;
